### ISSUE TITLE
Plotting processes have a chance to zombify when suspended.

### ISSUE THREAD

**Issue by Splitframe** [Chia-Network/chiapos#262]
### The Fault:
Suspending a plotting process has a chance to make it unresponsive. The longer the suspension the higher the chance the process will not respond to reanimation attempts. No Errors are given, it could be that it self-suspends immediately (?).

### The Problem:
Parallel plotting on drives that are fast enough, but lack the space to accommodate all plots being in P1/P2 require to temporarily suspend plots in those phases to ensure plots in P3/P4 can finish plotting.

### Free Text:
On 5 plotting machines I only encountered this problem on Windows so far.
It is aggravating though. Detecting the affected processes and restart plotting is not that big of an issue per se, but
the wasted time and resources do hurt.

**Comment by grayfallstown** [comment]
I encountered this issue as well a few times and thought it was just my setup.
I had a crashed chia.exe and leftover tempfiles from that. The next chia plotting process already started and the disk was almost full to the brim. 
I paused chia.exe via [Process Hacker](https://processhacker.sourceforge.io/) manually so I had time to cleanup temp files to free up space before the process crashes (back when write continuation did not work properly).

It happens quite rarely that a chia.exe process will not resume after being suspended. Twice it actually suddenly resumed. When the resume command was issued multiple times after a 30 minute lunch break the process did catch itself and resumed operation. I noticed that both the resume and the suspend option were available in Process Hacker at the same time which is usally not the case.

No idea how one would even try to approach this issue, tho.

**Comment by Splitframe** [comment]
To clarify, in my script I suspend/resume the processes via:

`def suspend_recursive(ps):
    for c in ps.children(recursive=True):
        c.suspend()
    ps.suspend()
    
def resume_recursive(ps):
    for c in ps.children(recursive=True):
        c.resume()
    ps.resume()`

**Comment by Splitframe** [comment]
I was able to narrow the problem down a little. So far I only encountered this on windows and only in Phase 1.
The zombified chia.exe has a python child and a python grandchild. Aggressively spamming resume on the grandchild seems to resolve this issue. For everyone making scripts I suggest to identify zombified processes and just loop spam them with resumes with a small timeout.

**Comment by arvidn** [comment]
it would be interesting to see what the process is doing when it's in this "zombie" state, if anyone could attach a debugger

**Comment by Splitframe** [comment]
```
Attaching to a process with PID=5,132
C:\<path>\python.exe "C:\Program Files\JetBrains\PyCharm 2020.3.3\plugins\python\helpers\pydev\pydevd_attach_to_process\attach_pydevd.py" --port 62415 --pid 5132
Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm 2020.3.3\plugins\python\helpers\pydev\pydevd_attach_to_process\attach_pydevd.py", line 86, in <module>
    main(process_command_line(sys.argv[1:]))
  File "C:\Program Files\JetBrains\PyCharm 2020.3.3\plugins\python\helpers\pydev\pydevd_attach_to_process\attach_pydevd.py", line 81, in main
    add_code_to_python_process.run_python_code(
  File "C:\Program Files\JetBrains\PyCharm 2020.3.3\plugins\python\helpers\pydev\pydevd_attach_to_process\add_code_to_python_process.py", line 306, in run_python_code_windows
    assert resolve_label(process, b'PyGILState_Ensure')
  File "C:\Program Files\JetBrains\PyCharm 2020.3.3\plugins\python\helpers\pydev\pydevd_attach_to_process\add_code_to_python_process.py", line 268, in resolve_label
    raise AssertionError('%s not resolved.' % (label,))
AssertionError: b'PyGILState_Ensure' not resolved.
```

Sorry I never attached a a debugger to a running process. I loaded chia-blockchain into pycharm, set the interpreter to the venv and tried to attach, but no cigar.

**Comment by arturzx** [comment]
> Sorry I never attached a a debugger to a running process. I loaded chia-blockchain into pycharm, set the interpreter to the venv and tried to attach, but no cigar.

Plotting is done in chiapos C++ code and attaching with Python debugger won't give you any information.

From stack overflow:
> From the Windows API perspective, there's no such thing as suspending a process. Only threads can be suspended, but there are no parent-child relationships between threads. Since there are no "child threads," there is no automatic mechanism for suspending them when the parent is suspended. (You can create a process suspended, but that's because when it's first created, there's only one thread, and it is created suspended.)

Maybe you must iterate threads and suspend it?

**Comment by Splitframe** [comment]
> Plotting is done in chiapos C++ code and attaching with Python debugger won't give you any information.

Yeah I know, but the really problematic process is python.exe, once I get this one running the others follow suit after some seconds and everything works again. So I thought I'd attach to this blocking python process.

I since solved the problem by just having a more elaborate and aggressive version of the resume function I posted earlier.
Somehow psutil seems to figure out a child-tree between processes/threads that works out for me.

In the Resource Manager this looks like this:
https://imgur.com/a/Bz1pxCc

In the first picture we see the wait chain of the chia.exe which has one of the paused python.exe, which in turn has the python.exe from the second picture in their wait chain. When I resume this last entry everything starts working again.
